File: validate_schemas.py
#!/usr/bin/env python3

import json
import jsonschema
from pathlib import Path
import sys

def validate_example(example_path, schema_path):
    """Validate an example JSON file against its schema"""
    try:
        # Load the schema and example
        with open(schema_path, 'r') as f:
            schema = json.load(f)

        with open(example_path, 'r') as f:
            example = json.load(f)

        # Validate
        jsonschema.validate(instance=example, schema=schema)
        print(f"✓ {example_path} validates against {schema_path}")
        return True

    except jsonschema.ValidationError as e:
        print(f"✗ {example_path} failed validation: {e.message}")
        return False
    except Exception as e:
        print(f"✗ Error validating {example_path}: {str(e)}")
        return False

def main():
    # Define schema-example pairs
    validations = [
        ("docs/schemas/examples/character_example.json", "docs/schemas/character.schema.json"),
        ("docs/schemas/examples/location_example.json", "docs/schemas/location.schema.json"),
        ("docs/schemas/examples/faction_example.json", "docs/schemas/faction.schema.json"),
        ("docs/schemas/examples/mechanics_example.json", "docs/schemas/mechanics_rule.schema.json"),
        ("docs/schemas/examples/story_example.json", "docs/schemas/story.schema.json")
    ]

    all_valid = True

    print("Validating examples against schemas...")
    print("=" * 50)

    for example_path, schema_path in validations:
        if not validate_example(example_path, schema_path):
            all_valid = False

    print("=" * 50)
    if all_valid:
        print("✓ All examples validated successfully!")
        return 0
    else:
        print("✗ Some examples failed validation")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: validate_schemas.js
#!/usr/bin/env node

const fs = require('fs');
const Ajv = require('ajv');
const ajv = new Ajv({ strict: false });

// Define schema-example pairs
const validations = [
    {
        example: 'docs/schemas/examples/character_example.json',
        schema: 'docs/schemas/character.schema.json',
        name: 'Character'
    },
    {
        example: 'docs/schemas/examples/location_example.json',
        schema: 'docs/schemas/location.schema.json',
        name: 'Location'
    },
    {
        example: 'docs/schemas/examples/faction_example.json',
        schema: 'docs/schemas/faction.schema.json',
        name: 'Faction'
    },
    {
        example: 'docs/schemas/examples/mechanics_example.json',
        schema: 'docs/schemas/mechanics_rule.schema.json',
        name: 'Mechanics Rule'
    },
    {
        example: 'docs/schemas/examples/story_example.json',
        schema: 'docs/schemas/story.schema.json',
        name: 'Story'
    }
];

console.log('Validating examples against schemas...');
console.log('='.repeat(50));

let allValid = true;

// Load and compile all schemas first
const schemas = {};
for (const validation of validations) {
    try {
        const schemaContent = fs.readFileSync(validation.schema, 'utf8');
        schemas[validation.schema] = JSON.parse(schemaContent);
        ajv.addSchema(schemas[validation.schema], validation.schema);
    } catch (error) {
        console.error(`✗ Failed to load schema ${validation.schema}: ${error.message}`);
        allValid = false;
    }
}

// Validate each example
for (const validation of validations) {
    try {
        const exampleContent = fs.readFileSync(validation.example, 'utf8');
        const exampleData = JSON.parse(exampleContent);

        const validate = ajv.getSchema(validation.schema);
        if (!validate) {
            console.error(`✗ No schema found for ${validation.schema}`);
            allValid = false;
            continue;
        }

        const valid = validate(exampleData);
        if (valid) {
            console.log(`✓ ${validation.name} example validates against schema`);
        } else {
            console.error(`✗ ${validation.name} example failed validation:`);
            if (validate.errors) {
                validate.errors.forEach(error => {
                    console.error(`  - ${error.instancePath || 'root'}: ${error.message}`);
                });
            }
            allValid = false;
        }
    } catch (error) {
        console.error(`✗ Error validating ${validation.example}: ${error.message}`);
        allValid = false;
    }
}

console.log('='.repeat(50));
if (allValid) {
    console.log('✓ All examples validated successfully!');
    process.exit(0);
} else {
    console.log('✗ Some examples failed validation');
    process.exit(1);
}